src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/AvoidMultipleEnumerations.DataFlowOperationVisitor.cs

```diff
@@ -37,52 +40,245 @@ public override GlobalFlowStateDictionaryAnalysisValue VisitLocalReference(ILoca
+            /// <summary>
+            /// Visit all the possible deferred type entities referenced by <param name="parameterOrLocalOperation"/>.
+            /// </summary>
+            private GlobalFlowStateDictionaryAnalysisValue VisitDeferTypeEntities(
+                PointsToAnalysisResult pointsToAnalysisResult,
+                IOperation parameterOrLocalOperation,
+                GlobalFlowStateDictionaryAnalysisValue defaultValue)
             {
-                var pointToAnalysisResult = DataFlowAnalysisContext.PointsToAnalysisResult;
-                invocationEntity = null;
-                if (pointToAnalysisResult is null)
+                RoslynDebug.Assert(parameterOrLocalOperation is IParameterReferenceOperation or ILocalReferenceOperation);
+                // With the initial operation as the root, expand it if
+                // the operation is parameter or local reference. It has only one pointToAnalysisResult, and it is a deferred execution invocation.
+                // Visit its argument.
+                // e.g.
+                // var a = b.Concat(c);
+                // a.ElementAt(10);
+                // When we visit 'a.Element(10)' and look back to 'b.Concat(c)', also try to visit 'b' and 'c'.
+                //
+                // Update the analysis value when reach one of the following nodes.
+                // 1. A parameter or local that has symbol, but no creationOperation.
+                // e.g.
+                // void Bar(IEnumerable<int> b, IEnumerable<int> c)
+                // {
+                //      var a = b.Concat(c);
+                //      a.ElementAt(10);
+                // }
+                // When 'a.ElementAt(10)' is called, 'b' and 'c' are enumerated once.
+                //
+                // 2. Invocation operation that returns a deferred type.
+                // e.g.
+                // void Bar()
+                // {
+                //       var a = Enumerable.Range(1, 1);
+                //       var b = Enumerable.Range(2, 2);
+                //       var c = a.Concat(b);
+                //       c.ElementAt(10);
+                // }
+                // When 'd.ElementAt(10)' is called, then 'Enumerable.Range(1, 1)', 'Enumerable.Range(2, 2)' and 'a.Concat(b)' are enumerated.
+                // 3. A parameter or local reference operation with multiple AbstractLocations. Stop expanding the tree at this node
+                // because we don't know how to proceed.
+                // e.g.
+                // void Bar(bool flag)
+                // {
+                //       var a = flag ? Enumerable.Range(1, 1) : Enumerable.Range(2, 2);
+                //       a.ElementAt(10);
+                // }
+                var queue = new Queue<IOperation>();
+                queue.Enqueue(parameterOrLocalOperation);
+                var resultAnalysisValue = defaultValue;
+                while (queue.Count > 0)
```

Comment:
A bit concerned about the perf implications here, we should measure and test perf with these changes.